### PR TITLE
repeatable ffdc support

### LINKED ISSUE #17815: Repeatable ExpectedFFDC and AllowedFFDC
We need this capability so tests (in saml sso fat projects and maybe some other tests) can be repeated and can have right actions go with the right exceptions.

### PR DESCRIPTION
**Pull request by arunavemulapalli**
fixes #17815

### CODE REVIEW

**Comment by arunavemulapalli** [comment]
#build

**Comment by arunavemulapalli** [comment]
#build

**Comment by arunavemulapalli** [comment]
#libby

**Comment by arunavemulapalli** [comment]
#libby

**Comment by arunavemulapalli** [comment]
#build

File: dev/com.ibm.ws.security.saml.sso_fat.2/fat/src/com/ibm/ws/security/saml/fat/common/BasicEncryptionTests.java [hunk 1]
```diff
@@ -332,7 +338,8 @@ public void testKeyAlias_MultipleCertsInKeystore_KeyAliasIsCorrectCert() throws
      * @throws Exception
      */
     @MaximumJavaLevel(javaLevel = 8) // test uses DSA cert and that is no longer supported
-    @AllowedFFDC(value = { "com.ibm.ws.security.saml.error.SamlException", "org.opensaml.xml.encryption.DecryptionException", "org.opensaml.xml.signature.SignatureException", "org.opensaml.xmlsec.encryption.support.DecryptionException", "org.opensaml.xmlsec.signature.support.SignatureException" })
+    @AllowedFFDC(value = { "com.ibm.ws.security.saml.error.SamlException", "org.opensaml.xml.encryption.DecryptionException" },repeatAction = {EmptyAction.ID})
```
Comment (jhanders34) [inline dev/com.ibm.ws.security.saml.sso_fat.2/fat/src/com/ibm/ws/security/saml/fat/common/BasicEncryptionTests.java]: Why is SignatureException no longer in either list?

File: dev/com.ibm.ws.security.saml.sso_fat.config/fat/src/com/ibm/ws/security/saml/sso/fat/config/SAMLMisc1ConfigTests.java [hunk 2]
```diff
@@ -900,7 +902,8 @@ public void test_config_loginPageURL_invalid() throws Exception {
      * time (not having to reconfig multiple times) we'll do all of this in one
      * test)
      */
-    @AllowedFFDC(value = { "com.ibm.ws.security.saml.error.SamlException", "org.opensaml.ws.security.SecurityPolicyException", "org.opensaml.xml.signature.SignatureException", "org.opensaml.messaging.handler.MessageHandlerException", "org.opensaml.xml.signature.SignatureException"  })
+    @AllowedFFDC(value = { "com.ibm.ws.security.saml.error.SamlException", "org.opensaml.ws.security.SecurityPolicyException", "org.opensaml.xml.signature.SignatureException" }, repeatAction = {EmptyAction.ID})
+    @AllowedFFDC(value = { "com.ibm.ws.security.saml.error.SamlException", "org.opensaml.messaging.handler.MessageHandlerException", "org.opensaml.xml.signature.SignatureException"  }, repeatAction = {JakartaEE9Action.ID})
```
Comment (jhanders34) [inline dev/com.ibm.ws.security.saml.sso_fat.config/fat/src/com/ibm/ws/security/saml/sso/fat/config/SAMLMisc1ConfigTests.java]: Should this one be `org.opensaml.xmlsec.signature.support.SignatureException` instead?

File: dev/fattest.simplicity/src/componenttest/annotation/AllowedFFDCs.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+package componenttest.annotation;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Retention(RUNTIME)
+@Target(METHOD)
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/annotation/AllowedFFDCs.java]: Why only method?

File: dev/fattest.simplicity/src/componenttest/annotation/ExpectedFFDCs.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+package componenttest.annotation;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Retention(RUNTIME)
+@Target(METHOD)
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/annotation/ExpectedFFDCs.java:20]: Why only method?

File: dev/fattest.simplicity/src/componenttest/custom/junit/runner/FATRunner.java [hunk 5]
```diff
@@ -664,13 +666,13 @@ private String getFFDCHeader(RemoteFile ffdcLogFile) throws Exception {
     private Set<String> getAllowedFFDCAnnotationFromTest(FrameworkMethod m) {
 
         Set<String> annotationListPerClass = new HashSet<String>();
-
         // Method
-        Set<AllowedFFDC> ffdcs = new HashSet<AllowedFFDC>();
-        ffdcs.add(m.getAnnotation(AllowedFFDC.class));
+        AllowedFFDC[] allowedffdcs = m.getMethod().getAnnotationsByType(AllowedFFDC.class);
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/custom/junit/runner/FATRunner.java:670]: Why isn't the logic here the same as ExpectedFFDC?

**Review by jhanders34 — CHANGES_REQUESTED** [review]

File: dev/fattest.simplicity/src/componenttest/annotation/ExpectedFFDCs.java (hunk 4 above)
Comment (arunavemulapalli) [reply dev/fattest.simplicity/src/componenttest/annotation/ExpectedFFDCs.java:20]: To match the ExpectedFFDC

File: dev/fattest.simplicity/src/componenttest/custom/junit/runner/FATRunner.java (hunk 5 above)
Comment (arunavemulapalli) [reply dev/fattest.simplicity/src/componenttest/custom/junit/runner/FATRunner.java:670]: The difference between the original ExpectedFFDC (which is restricted to @Target({ ElementType.METHOD }))and AllowedFFDC annotations is the reason for the difference in logic here and I did not alter it.

File: dev/com.ibm.ws.security.saml.sso_fat.2/fat/src/com/ibm/ws/security/saml/fat/common/BasicEncryptionTests.java (hunk 1 above)
Comment (arunavemulapalli) [reply dev/com.ibm.ws.security.saml.sso_fat.2/fat/src/com/ibm/ws/security/saml/fat/common/BasicEncryptionTests.java]: fixed it.

File: dev/com.ibm.ws.security.saml.sso_fat.config/fat/src/com/ibm/ws/security/saml/sso/fat/config/SAMLMisc1ConfigTests.java (hunk 2 above)
Comment (arunavemulapalli) [reply dev/com.ibm.ws.security.saml.sso_fat.config/fat/src/com/ibm/ws/security/saml/sso/fat/config/SAMLMisc1ConfigTests.java]: yes, fixed it.

File: dev/fattest.simplicity/src/componenttest/annotation/AllowedFFDCs.java (hunk 3 above)
Comment (arunavemulapalli) [reply dev/fattest.simplicity/src/componenttest/annotation/AllowedFFDCs.java]: removed it. AllowedFFDC is not restricted.

**Comment by arunavemulapalli** [comment]
#libby

**Comment by arunavemulapalli** [comment]
#build

**Comment by arunavemulapalli** [comment]
#build

**Review by jhanders34 — APPROVED** [review]

**Pull request merged**